share/shwap/proof.go

```diff
@@ -0,0 +1,220 @@
+package shwap
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/tendermint/tendermint/crypto/merkle"
+
+	libshare "github.com/celestiaorg/go-square/v2/share"
+	"github.com/celestiaorg/nmt"
+	nmt_pb "github.com/celestiaorg/nmt/pb"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/shwap/pb"
+)
+
+// Proof represents a proof that a data share is included in a
+// committed data root.
+// It consists of multiple components to support verification:
+// * shareProof: A nmt proof that verifies the inclusion of a specific data share within a row of the datasquare.
+// * rowRootProof: A Merkle proof that verifies the inclusion of the row root within the final data root.
+// * root: The row root against which the proof is verified
+type Proof struct {
+	shareProof   *nmt.Proof
+	rowRootProof *merkle.Proof
+	root         []byte
+}
```

Comment:
This proof currently uses a two-step inclusion verification:

- Inclusion of a row into the row root.
- Inclusion of row roots into the data root.

This proof structure isn't optimal because a range request would require multiple row-root proofs, which is unnecessary. It should be possible to optimize proofs for `GetRange` over shwap, as we already have the necessary code implemented in `types.Proof` (from the app repo).

It's important to address this now to avoid introducing breaking changes in a future P2P release, which would otherwise require a network migration and significant additional effort.